enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -2257,30 +2250,61 @@ func (c *FirecrackerContainer) Exec(ctx context.Context, cmd *repb.Command, stdi
 		c.workspaceStore.EmitUsageMetrics("init")
 	}
 
-	result = c.SendExecRequestToGuest(ctx, cmd, workDir, stdio)
+	// TODO(bduffany): Reuse connection from Unpause(), if applicable
+	conn, err := c.dialVMExecServer(ctx)
+	if err != nil {
+		return commandutil.ErrorResult(status.InternalErrorf("Firecracker exec failed: failed to dial VM exec port: %s", err))
+	}
+	defer conn.Close()
+
+	result = c.SendExecRequestToGuest(ctx, conn, cmd, workDir, stdio)
 	close(execDone)
 
 	ctx, cancel = background.ExtendContextForFinalization(ctx, finalizationTimeout)
 	defer cancel()
 
 	// If FUSE is enabled then outputs are already in the workspace.
 	if c.fsLayout == nil && !*disableWorkspaceSync {
-		// Command was successful, let's unpack the files back to our
-		// workspace directory now.
+		// Unmount the workspace and pause the VM before syncing to ensure that
+		// the guest's FS cache is flushed to the backing file on the host and
+		// that no concurrent operations can occur while we're reading the
+		// workspace disk.
+		//
+		// TODO(bduffany): after notifying users of these warnings below,
+		// make it a hard failure if we fail to unmount or if the workspace disk
+		// is still busy after unmounting.
```

Comment:
The logic here is a bit ugly but the idea is to avoid breaking actions which are working today just because this new Unmount logic is failing. Once this is rolled out I would like to check and see how often these logs are happening in practice and then make the unmount failures real errors instead of warnings.